File: app/src/main/java/com/fastaccess/ui/adapter/MilestonesAdapter.java
package com.fastaccess.ui.adapter;

import android.support.annotation.NonNull;
import android.view.ViewGroup;

import com.fastaccess.data.dao.MilestoneModel;
import com.fastaccess.ui.adapter.viewholder.MilestonesViewHolder;
import com.fastaccess.ui.widgets.recyclerview.BaseRecyclerAdapter;
import com.fastaccess.ui.widgets.recyclerview.BaseViewHolder;

import java.util.ArrayList;

/**
 * Created by Kosh on 11 Nov 2016, 2:07 PM
 */

public class MilestonesAdapter extends BaseRecyclerAdapter<MilestoneModel, MilestonesViewHolder,
        BaseViewHolder.OnItemClickListener<MilestoneModel>> {

    public MilestonesAdapter(@NonNull ArrayList<MilestoneModel> eventsModels) {
        super(eventsModels);
    }

    @Override protected MilestonesViewHolder viewHolder(ViewGroup parent, int viewType) {
        return MilestonesViewHolder.newInstance(parent, this);
    }

    @Override protected void onBindView(MilestonesViewHolder holder, int position) {
        holder.bind(getItem(position));
    }
}

File: app/src/main/java/com/fastaccess/github/ui/adapter/MilestonesAdapter.kt
package com.fastaccess.github.ui.adapter

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.fastaccess.data.model.parcelable.MilestoneModel
import com.fastaccess.github.ui.adapter.viewholder.MilestoneViewHolder

/**
 * Created by Kosh on 20.01.19.
 */
class MilestonesAdapter(
    private val clickCallback: (MilestoneModel) -> Unit
) : ListAdapter<MilestoneModel, MilestoneViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MilestoneViewHolder = MilestoneViewHolder(parent).apply {
        itemView.let { view ->
            view.setOnClickListener {
                getItem(adapterPosition)?.let(clickCallback)
            }
        }
    }

    override fun onBindViewHolder(holder: MilestoneViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item)
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<MilestoneModel?>() {
            override fun areItemsTheSame(oldItem: MilestoneModel, newItem: MilestoneModel): Boolean = oldItem.id == newItem.id
            override fun areContentsTheSame(oldItem: MilestoneModel, newItem: MilestoneModel): Boolean = oldItem == newItem
        }
    }
}
